clustermesh: Consolidate hive

joestringer opened the pull request:
Since merging clustermesh and kvstoremesh together, the code was still
quite separate, registering multiple Hives and duplicate Cells (+
duplicate configuration). This could cause some options to be not
responsive to user configuration, as well as just unnecessarily running
extra instances of core logic that is not necessary to run - for
instance, two pprof or two gops servers. This may also fix issues where
some clustermesh controllers do not appear in debug status output.

Consolidate these into a single Hive with all the common cells, and
expose a new 'clustermesh-apiserver hive' command to inspect the
hierarchy.

giorio94 commented on Documentation/cmdref/clustermesh-apiserver_kvstoremesh.md:
```diff
@@ -16,10 +16,11 @@ clustermesh-apiserver kvstoremesh [flags]
       --cluster-name string                          Name of the cluster. It must consist of at most 32 lower case alphanumeric characters and '-', start and end with an alphanumeric character. (default "default")
       --clustermesh-config string                    Path to the ClusterMesh configuration directory
       --controller-group-metrics strings             List of controller group names for which to to enable metrics. Accepts 'all' and 'none'. The set of controller group names available is not guaranteed to be stable between Cilium versions.
+      --crd-wait-timeout duration                    Cilium will exit if CRDs are not available within this duration upon startup (default 5m0s)
```
This flag is not applicable to KVStoreMesh, and should not be registered here.

joestringer replied:
Sorry, yeah I should have caught this during my own self-review. Thanks for the detailed review, I'll address your feedback.

giorio94 commented on clustermesh-apiserver/clustermesh/cells.go:
```diff
@@ -7,54 +7,28 @@ import (
 	// case we only use the external workloads feature, and not clustermesh.
-	cell.Config(cmtypes.DefaultClusterInfo),
 	cell.Invoke(cmtypes.ClusterInfo.InitClusterIDMax),
 	cell.Invoke(cmtypes.ClusterInfo.Validate),
```
Just a note that this part will be uniformed with KVStoreMesh via https://github.com/cilium/cilium/pull/37418

giorio94 commented on clustermesh-apiserver/health/health.go:
```diff
@@ -71,3 +76,31 @@ func registerHealthAPIServer(lc cell.Lifecycle, params parameters) {
+	Logger    logrus.FieldLogger
+}
+
+func healthEndpoints(params healthParameters) []EndpointFunc {
```
I'm personally a bit on the fence on the opportunity of merging this function, instead of keeping it separate for each component. The rationale for having each component defining its own `healthEndpoints` is that the checks may differ (e.g., the clustermesh-apiserver may additionally check connectivity to Kubernetes, which does not make sense for KVStoreMesh). Hence, having separate functions would make extending the checks easier. That said, I also reckon that the two functions are identical at the moment.

joestringer replied:
It logically makes sense for them to be different, so I split this such that the API server itself is common but the health endpoint declarations like this remain split. That should help with any future changes to extend these APIs.

giorio94 commented on clustermesh-apiserver/main.go:
```diff
@@ -7,16 +7,52 @@ import (
+	"github.com/cilium/cilium/pkg/pprof"
+)
+
+var commonCell = cell.Module(
```
Could you please move the definition of the common cell to a dedicated package, so that it is easier to extend in the future without polluting the main package?

giorio94 commented on Documentation/cmdref/clustermesh-apiserver_kvstoremesh.md:
```diff
@@ -16,10 +16,11 @@ clustermesh-apiserver kvstoremesh [flags]
       --enable-gops                                  Enable gops server (default true)
       --global-ready-timeout duration                KVStoreMesh will be considered ready even if any remote clusters have failed to synchronize within this duration (default 10m0s)
-      --gops-port uint16                             Port for gops server to listen on (default 9894)
+      --gops-port uint16                             Port for gops server to listen on (default 9892)
```
This change should be reverted, because the default port used by ClusterMesh needs to be different compared to the one used by KVStoreMesh. Otherwise, both processes would listen to the same port (which apparently doesn't lead to an explicit error because [ReuseSocketAddrAndPort is set](https://github.com/cilium/cilium/blob/pr%2Fjoe%2Fclustermesh-hive/pkg/gops/cell.go#L50-L53)), hence leading to unexpected behavior.

joestringer replied:
Understood. I wonder if I can find a way to parameterize these, or we'll just retain the duplicate definitions.

giorio94 replied:
One thing I did in the past to parametrize a cell was making it a function that takes the configuration as parameters, like in https://github.com/cilium/cilium/commit/2911b44ace7b4749ae3725d5f74be23d7f792fc9#diff-468f812f6f539a7a06cb9e901a1d4c9ab3ce073801788680ac3dbb4f91a7992aR27. Otherwise, I don't think there are many other alternatives than moving the definition of the configuration outside of the common cell.

giorio94 commented on Documentation/cmdref/clustermesh-apiserver_kvstoremesh.md:
```diff
@@ -34,7 +35,7 @@ clustermesh-apiserver kvstoremesh [flags]
       --pprof                                        Enable serving pprof debugging API
       --pprof-address string                         Address that pprof listens on (default "localhost")
-      --pprof-port uint16                            Port that pprof listens on (default 6064)
+      --pprof-port uint16                            Port that pprof listens on (default 6063)
```
Same here.

giorio94 reviewed (changes requested):
Thanks for taking care of this. A couple of comments inline.

joestringer committed:
pprof: Parameterize config

There are several components which rely on the pprof Cell, with
different configuration parameters for each. In order to improve code
reuse, require the pprof configuration as a parameter for the Cell and
expect the caller to define the default configuration for those flags.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
synced: Move CRDWaitTimeout to Cell Flag

This flag was previously declared in two places for the agent and
clustermesh-apiserver to define the configuration, however it's used
only from this synced package. Move the flag configuration into this
package to consolidate the declaration into one place for better code
reuse.

The agent hive command reference here presumably changes because the
flag was migrated from the traditional global flags reference over into
the Hive, so the Hive infrastructure is now aware of the flag.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
k8s: Remove unused function WaitForCRDsToRegister()

This function has been unused since the introduction of the CRD sync
promise. Remove it as it is no longer needed.

Fixes: d8102b98f04b ("k8s/synced: Add CRDSync promise")
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
clustermesh-apiserver: De-duplicate legacy config

These "legacy" configurations are common across both
clustermesh/kvstoremesh binaries, so we can reuse the same flag
declarations for both. While we're at it, explicitly define the default
configuration for the log options.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
clustermesh: Move store.Cell declaration

An upcoming change will move this into a common package. In preparation,
move this out to the top level so that the diff in the subsequent commit
is clearer to follow. No logical changes here.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
clustermesh: Consolidate hive

Several of the core Cells that are reused by clustermesh and kvstoremesh
are declared separately. In order to simplify the logic, this commit
moves those common Cells into a single declaration at the top level and
directly passes that common Cell into each Hive when it is created on
startup.

The gops,pprof cells are kept separately declared in order to retain the
different port configuration to avoid conflicts when running on the same
IP. The Health API server is centralized, but the health endpoints APIs
are kept separate in order to allow them to diverge in future (though
they are currently identical).

Signed-off-by: Joe Stringer <joe@cilium.io>

giorio94 reviewed (approved):
Thanks!

dylandreimerink reviewed (approved):
I think this is a nice change. Can't find anything wrong with this
